jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3439,31 +3440,60 @@ public static MethodHandle whileLoop(MethodHandle initHandle, MethodHandle predH
+		 * The constraints of initializer and predicate at other aspects will be checked
 		 * later in the generic loop when adding clauses.
 		 */
-		MethodHandle finiHandle = validateArgumentsOfWhileLoop(predHandle, bodyHandle);
+		MethodHandle finiHandle = validateArgumentsOfWhileLoop(initHandle, predHandle, bodyHandle);
 		
 		/* Create the clause array with all these handles to be used by the generic loop */
 		MethodHandle[] handleClause = {initHandle, bodyHandle, predHandle, finiHandle};
 		return loop(handleClause);
 	}
 	
 	/* Validate the body handle against the constraints of a while/do-while loop and generate the 'fini' handle from it */
-	private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle predHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
+	private static MethodHandle validateArgumentsOfWhileLoop(MethodHandle initHandle, MethodHandle predHandle, MethodHandle bodyHandle) throws NullPointerException, IllegalArgumentException {
 		if ((null == predHandle) || (null == bodyHandle)) {
 			/*[MSG "K065C", "Both The predicate and the loop body must be non-null"]*/
 			throw new NullPointerException(com.ibm.oti.util.Msg.getString("K065C")); //$NON-NLS-1$
 		}
 		
 		MethodType bodyType = bodyHandle.type;
 		Class<?> bodyReturnType = bodyType.returnType;
+		int bodyParamLength = bodyType.parameterCount();
 		
 		/* The signature of the body handle must be either (V, A...)V or (A...)void */
-		if ((void.class != bodyReturnType) && (bodyReturnType != bodyType.arguments[0])) {
-			/*[MSG "K065D", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
-			throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D", bodyReturnType.getName(), bodyType.arguments[0].getName())); //$NON-NLS-1$
+		int bodyIterationVarLength = 0;
+		if (void.class != bodyReturnType) {
+			if (hasNoArgs(bodyType)) {
+				/*[MSG "K065D3", "loop body has a non-void return type and must contain at least one argument. First argument must be {0}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D3", bodyReturnType)); //$NON-NLS-1$
+			}
+			
+			if (bodyReturnType != bodyType.arguments[0]) {
+				/*[MSG "K065D2", "The return type of loop body: {0} does not match the type of the first argument: {1}"]*/
+				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065D2", bodyReturnType, bodyType.arguments[0])); //$NON-NLS-1$
+			}
+			
+			bodyIterationVarLength = 1;
+		}
+		
+		/* Verify that handles are effectively identical and constrained by body. All we
+		 * need to do here is verify that no parameters have longer argument lists than
+		 * body. The generic loop will compare arguments to ensure all clauses are
+		 * effectively identical.
+		 */
+		int bodyExternalParamLength = bodyParamLength - bodyIterationVarLength;
+		
+		if (predHandle.type.parameterCount() > bodyParamLength) {
```

Comment:
Is this mean to be `bodyParamLength` or `bodyExternalParamLength`?  The comment error message refers to "external parameters"

Reply:
If it's not, can the calculation of `bodyExternalParamLength` be moved closer to its use?

Reply:
Logically the code here is correct as there is only 1 iteration variable (type V) for predicate(V, A...) and body(V, A...).  (bodyIterationVarLength = 1).  Actually it is comparing is the A list of predicate with the A list of body.

But the error message is misleading and easier to get reader confused about the code against the message.

@theresa-m,  one option is to change it  to:
 if ((predHandle.type.parameterCount() - bodyIterationVarLength) > bodyExternalParamLength) {
...

Reply:
Thanks @ChengJin01 . I've updated the error message for the pred handle since it doesn't make much sense and added a comment about why this case is different.